### ISSUE TITLE
Azure Cache Redis is not having (is empty of) data sometimes

### ISSUE THREAD

**Issue by cosmycx** [CDCgov/data-exchange-hl7#432]
**Describe the bug**
Missing all data in Redis
![image](https://user-images.githubusercontent.com/11874351/208922642-c314714e-45c3-407e-ac54-e774134d7472.png)

**Impact**
- Pipeline is broke.
- Local functions are broke. 
![image](https://user-images.githubusercontent.com/11874351/208923224-2c546e1b-6e73-44e2-8102-d388fb8dd000.png)

**Comment by mscaldas2012** [comment]
It seems that Redis gets patched every Wed. at 5am
![image](https://user-images.githubusercontent.com/3239945/208925670-d3c10c70-94b0-4c55-bd92-fd7ca9221a83.png)

**Comment by mscaldas2012** [comment]
I've created an Alert that monitors if Redis data deeps below 100Mb, it will send a Critical Alert (email me, boris and Neelima - previously set)
![image](https://user-images.githubusercontent.com/3239945/208926007-459b718f-9832-47dc-8adb-e947df4d43d1.png)

**Comment by mscaldas2012** [comment]
It seems we can attach an event to Azure being patched. but our functions (schedule trigger) are not compatible
![image](https://user-images.githubusercontent.com/3239945/208927470-d6be05ce-0656-45d4-943c-288a9956d45d.png)

**Comment by mscaldas2012** [comment]
There's also ways of automating Export/Import data from/to Redis:
https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-how-to-import-export-data

**Comment by mscaldas2012** [comment]
TO use our Loader functions, we need to add either an EventGrid or HTTP trigger:

Use [Event Grid trigger](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-grid-trigger). Specify Azure Function as the endpoint type. Then, specify the function app and the function that will handle events.
Use [HTTP trigger](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook). Specify Web Hook as the endpoint type. Then, specify the URL for the function that will handle events.

(from https://learn.microsoft.com/en-us/azure/event-grid/handler-functions)

**Comment by rmharrison** [comment]
Done
- Email alert when available space on Redis drops below 100Mb (on ~1GB total)

Resolutions for PROD
- Option A: Upgrade to Redis Premium so we have a "no downtime" SLA (99.99...%)
- Option B: Scheduled downtown. Turn off the pipeline before Redis outage; turn it back on after patch

**Comment by mscaldas2012** [comment]
We identified that we can also trigger a HTTP function from the Alert itself. so when the alert fires (Redis mem < 100MB), the Function can run 

Alert Rules -> Action groups (potential solution) to call function app to refill Redis.

Once this is implemented and tested, we can ask Boris to Terraform this solution

**Comment by mscaldas2012** [comment]
Ticket #526 created to handle the resolution

**Comment by cosmycx** [comment]
The following is for OPTION 2 above, which as this time is the recommended solution, below details:

TLDR: Change Redis to Premium P1, use 2 instances and share: dev + tst and prod + stg

### Redis  - Azure Cache Redis
- solution 1: Basic C1 (currently using)
- solution 2: Premium P1

| Cache Name  | Cache Size  |  Network performance  | Number of Client Connections  |  Price       |  1 YR Reserved           |  3 YR Reserved           |
| ----------- | ----------- |  -----------          |  -----------                  |  ----------- |  -----------             |  -----------             |
| C1          | 1 GB        | Low                   | 1,000                         | $0.055/hour  |                          |                          |
| P1          | 6 GB        | Moderate              | 7,500                         | $0.554/hour  | $0.355/hour ~36% savings | $0.250/hour ~55% savings |


### Basic C1
- The default maintenance window is 5 hours 
- Single node cache is a single node cache which is ideal for development/test and non-critical workloads
- There's no SLA on basic tier
  
This Basic C1 would require custom coded solution to accomplish pesistence, meaning added initial cost and life cycle cost.
Multiple DEX Projects (such as HL7, etc) sharing Basic C1 is not feasible now due to now multiple custom coded solutions, coordination of those, and see above.

### Premium 
- Premium tier includes all the standard-tier features and more, such as better performance over basic or standard-tier caches, bigger workloads, disaster recovery, and enhanced security
  - Persistence
  - Cluster
  - AVN azure virtual network 
  - Multi-replica support

The Premium P1 is recommended for Prod., it could be shared between prod. and stg.
Hence a Premium P1 should be used for Dev. and Tst.

- Dev. and Tst. could eventually switch back, once development is not active to Basic
- Premium P1 could be 1-3 Yrs. reserved - this is a larger infra issue that should account other services for reservation and cost saving. (once dev. and full solutioning is more complete.)

Ref:
Azure cache redis pricing: https://azure.microsoft.com/en-us/pricing/details/cache/
Redis persistence: https://redis.io/docs/management/persistence/

**Comment by cosmycx** [comment]
@mscaldas2012  @boris-ning-usds - we are leaning strongly on switching to Az Redis Premium, would you please check that setting persistence on it is an easy check (or simple az process). please see above comment details on why the switch and we can discuss more details or questions if you have. Thanks
